[FLINK-29776][flink-statebackend-changelog][JUnit5 Migration]Module: flink-statebackend-changelog.

jiexray opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

[JUnit5 Migration] Module: flink-statebackend-changelog


## Brief change log

  - *[JUnit5 Migration] Module: flink-statebackend-changelog
  - *[Junit5 Migration] partial unit test base class in flink-runtime and flink-statebackend-rocksdb, i.e., `StateBackendMigrationTestBase`, `StateBackendTestBase`, and their child ut classes.


## Verifying this change

This change is already covered by existing tests, such as (please describe tests).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

jiexray commented:
Hello @PatrickRen @ruanhang1993, I am trying to migrate junit tests in flink-statebackend-changelog. Some base tests for changelog are in flink-runtime. The two base test classes are `StateBackendTestBase` and `StateBackendMigrationTestBase`. Could you help me to resolve the conflict.

jiexray commented:
Hi, @piotrp @Myasuka Would you mind helping to check this flink-statebackend-changelog JUnit5 Migration ? Thank you very much.

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -74,6 +75,7 @@
  */
 @SuppressWarnings("serial")
+@ExtendWith(ParameterizedTestExtension.class)
 public abstract class StateBackendMigrationTestBase<B extends StateBackend> extends TestLogger {
```
`TestLogger` is a JUnit4 rule/extension
Here should be used `META-INF/services/org.junit.jupiter.api.extension.Extension`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -98,14 +100,14 @@ protected boolean supportsKeySerializerCheck() {
 
-    @Before
+    @BeforeEach
     public void before() {
```
```suggestion
    void before() {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -98,14 +100,14 @@ protected boolean supportsKeySerializerCheck() {
 
-    @After
+    @AfterEach
     public void after() {
```
```suggestion
    void after() {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -119,7 +121,7 @@ public void after() {
 
-    @Test
+    @TestTemplate
     public void testKeyedValueStateMigration() throws Exception {
```
```suggestion
    void testKeyedValueStateMigration() throws Exception {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -131,7 +133,7 @@ public void testKeyedValueStateMigration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedValueStateSerializerReconfiguration() throws Exception {
```
```suggestion
    void testKeyedValueStateSerializerReconfiguration() throws Exception {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -144,7 +146,7 @@ public void testKeyedValueStateSerializerReconfiguration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedValueStateRegistrationFailsIfNewStateSerializerIsIncompatible() {
```
```suggestion
    void testKeyedValueStateRegistrationFailsIfNewStateSerializerIsIncompatible() {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -156,9 +158,10 @@ public void testKeyedValueStateRegistrationFailsIfNewStateSerializerIsIncompatib
 
             fail("should have failed");
         } catch (Exception expected) {
-            Assert.assertTrue(
-                    ExceptionUtils.findThrowable(expected, StateMigrationException.class)
-                            .isPresent());
+            assertThat(
+                            ExceptionUtils.findThrowable(expected, StateMigrationException.class)
+                                    .isPresent())
+                    .isTrue();
```
I guess this could be replaced with usage of `assertThatThrownBy`
```java
 assertThatThrownBy(
                        () ->
                                testKeyedValueStateUpgrade(
                                        new ValueStateDescriptor<>(
                                                stateName, new TestType.V1TestTypeSerializer()),
                                        new ValueStateDescriptor<>(
                                                stateName,
                                                new TestType.IncompatibleTestTypeSerializer())))
                .isInstanceOf(StateMigrationException.class);
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -237,7 +240,7 @@ private void testKeyedValueStateUpgrade(
 
-    @Test
+    @TestTemplate
     public void testKeyedListStateMigration() throws Exception {
```
```suggestion
    void testKeyedListStateMigration() throws Exception {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -249,7 +252,7 @@ public void testKeyedListStateMigration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedListStateSerializerReconfiguration() throws Exception {
```
```suggestion
    void testKeyedListStateSerializerReconfiguration() throws Exception {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -262,7 +265,7 @@ public void testKeyedListStateSerializerReconfiguration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedListStateRegistrationFailsIfNewStateSerializerIsIncompatible() {
```
```suggestion
    void testKeyedListStateRegistrationFailsIfNewStateSerializerIsIncompatible() {
```
Could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -274,9 +277,10 @@ public void testKeyedListStateRegistrationFailsIfNewStateSerializerIsIncompatibl
         } catch (Exception expected) {
-            Assert.assertTrue(
-                    ExceptionUtils.findThrowable(expected, StateMigrationException.class)
-                            .isPresent());
+            assertThat(
+                            ExceptionUtils.findThrowable(expected, StateMigrationException.class)
+                                    .isPresent())
+                    .isTrue();
```
i guess it could be replaced with usage of `assertThatThrownBy`
something like
```java
        assertThatThrownBy(
                        () ->
                                testKeyedListStateUpgrade(
                                        new ListStateDescriptor<>(
                                                stateName, new TestType.V1TestTypeSerializer()),
                                        new ListStateDescriptor<>(
                                                stateName,
                                                new TestType.IncompatibleTestTypeSerializer())))
                .isInstanceOf(StateMigrationException.class);
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -370,7 +374,7 @@ private void testKeyedListStateUpgrade(
 
-    @Test
+    @TestTemplate
     public void testKeyedMapStateAsIs() throws Exception {
```
```suggestion
    void testKeyedMapStateAsIs() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -381,7 +385,7 @@ public void testKeyedMapStateAsIs() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedMapStateStateMigration() throws Exception {
```
```suggestion
    void testKeyedMapStateStateMigration() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -395,7 +399,7 @@ public void testKeyedMapStateStateMigration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedMapStateSerializerReconfiguration() throws Exception {
```
```suggestion
    void testKeyedMapStateSerializerReconfiguration() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -409,7 +413,7 @@ public void testKeyedMapStateSerializerReconfiguration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyedMapStateRegistrationFailsIfNewStateSerializerIsIncompatible() {
```
```suggestion
    void testKeyedMapStateRegistrationFailsIfNewStateSerializerIsIncompatible() {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -424,9 +428,10 @@ public void testKeyedMapStateRegistrationFailsIfNewStateSerializerIsIncompatible
                             new TestType.IncompatibleTestTypeSerializer()));
             fail("should have failed");
         } catch (Exception expected) {
-            Assert.assertTrue(
-                    ExceptionUtils.findThrowable(expected, StateMigrationException.class)
-                            .isPresent());
+            assertThat(
+                            ExceptionUtils.findThrowable(expected, StateMigrationException.class)
+                                    .isPresent())
+                    .isTrue();
```
I guess it could be replaced with usage of `assertThatThrownBy`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -572,8 +577,8 @@ public void testPriorityQueueStateCreationFailsIfNewSerializerIsNotCompatible()
 
             fail("should have failed");
         } catch (Exception e) {
-            Assert.assertTrue(
-                    ExceptionUtils.findThrowable(e, StateMigrationException.class).isPresent());
+            assertThat(ExceptionUtils.findThrowable(e, StateMigrationException.class).isPresent())
+                    .isTrue();
```
I guess it could be replaced with usage of `assertThatThrownBy`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -583,7 +588,7 @@ public void testPriorityQueueStateCreationFailsIfNewSerializerIsNotCompatible()
 
-    @Test
+    @TestTemplate
     public void testStateBackendRestoreFailsIfNewKeySerializerRequiresMigration() throws Exception {
```
```suggestion
    void testStateBackendRestoreFailsIfNewKeySerializerRequiresMigration() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -593,13 +598,14 @@ public void testStateBackendRestoreFailsIfNewKeySerializerRequiresMigration() th
 
-    @Test
+    @TestTemplate
     public void testStateBackendRestoreSucceedsIfNewKeySerializerRequiresReconfiguration()
```
```suggestion
    void testStateBackendRestoreSucceedsIfNewKeySerializerRequiresReconfiguration()
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -1052,7 +1061,7 @@ public void testBroadcastStateKeySerializerReconfiguration() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testBroadcastStateRegistrationFailsIfNewValueSerializerIsIncompatible() {
```
```suggestion
    void testBroadcastStateRegistrationFailsIfNewValueSerializerIsIncompatible() {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -1195,12 +1204,11 @@ public void internalCopySerializerTest(
 
-    @Test
+    @TestTemplate
     public void testStateMigrationAfterChangingTTL() throws Exception {
```
```suggestion
    void testStateMigrationAfterChangingTTL() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -1216,7 +1224,7 @@ public void testStateMigrationAfterChangingTTL() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testStateMigrationAfterChangingTTLFromEnablingToDisabling() {
```
```suggestion
    void testStateMigrationAfterChangingTTLFromEnablingToDisabling() {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -1231,13 +1239,14 @@ public void testStateMigrationAfterChangingTTLFromEnablingToDisabling() {
 
-    @Test
+    @TestTemplate
     public void testStateMigrationAfterChangingTTLFromDisablingToEnabling() {
```
```suggestion
    void testStateMigrationAfterChangingTTLFromDisablingToEnabling() {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -1231,13 +1239,14 @@ public void testStateMigrationAfterChangingTTLFromEnablingToDisabling() {
             fail("should have failed");
         } catch (Exception expected) {
-            Assert.assertTrue(
-                    ExceptionUtils.findThrowable(expected, StateMigrationException.class)
-                            .isPresent());
+            assertThat(
+                            ExceptionUtils.findThrowable(expected, StateMigrationException.class)
+                                    .isPresent())
+                    .isTrue();
```
I guess it could be replaced with usage of `assertThatThrownBy`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendMigrationTestBase.java:
```diff
@@ -1253,9 +1262,10 @@ public void testStateMigrationAfterChangingTTLFromDisablingToEnabling() {
             fail("should have failed");
         } catch (Exception expected) {
-            Assert.assertTrue(
-                    ExceptionUtils.findThrowable(expected, IllegalStateException.class)
-                            .isPresent());
+            assertThat(
+                            ExceptionUtils.findThrowable(expected, IllegalStateException.class)
+                                    .isPresent())
+                    .isTrue();
```
I guess it could be replaced with usage of `assertThatThrownBy`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -152,16 +137,15 @@
-
-    @Before
+    @BeforeEach
     public void before() throws Exception {
```
```suggestion
    void before() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -152,16 +137,15 @@
 
-    @After
+    @AfterEach
     public void after() {
```
```suggestion
    void after() {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -152,16 +137,15 @@
  */
 @SuppressWarnings("serial")
+@ExtendWith(ParameterizedTestExtension.class)
 public abstract class StateBackendTestBase<B extends AbstractStateBackend> extends TestLogger {
```
`TestLogger` is a JUnit4 rule/extension
Here should be used `META-INF/services/org.junit.jupiter.api.extension.Extension`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -276,7 +260,7 @@ protected <K> CheckpointableKeyedStateBackend<K> restoreKeyedBackend(
 
-    @Test
+    @TestTemplate
     public void testEnableStateLatencyTracking() throws Exception {
```
```suggestion
    void testEnableStateLatencyTracking() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -303,34 +287,35 @@ public void testEnableStateLatencyTracking() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testIsSafeToReuseState() throws Exception {
```
```suggestion
    void testIsSafeToReuseState() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -303,34 +287,35 @@ public void testEnableStateLatencyTracking() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyGroupedInternalPriorityQueue() throws Exception {
```
```suggestion
    void testKeyGroupedInternalPriorityQueue() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -303,34 +287,35 @@ public void testEnableStateLatencyTracking() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testKeyGroupedInternalPriorityQueueAddAll() throws Exception {
```
```suggestion
    void testKeyGroupedInternalPriorityQueueAddAll() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -355,43 +340,41 @@ public void testKeyGroupedInternalPriorityQueue(boolean addAll) throws Exception
 
-    @Test
+    @TestTemplate
     public void testGetKeys() throws Exception {
```
```suggestion
    void testGetKeys() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -446,19 +429,19 @@ public void testGetKeys() throws Exception {
 
-    @Test
+    @TestTemplate
     public void testGetKeysAndNamespaces() throws Exception {
```
```suggestion
    void testGetKeysAndNamespaces() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -506,7 +491,7 @@ public void testGetKeysAndNamespaces() throws Exception {
-    @Test
+    @TestTemplate
     @SuppressWarnings("unchecked")
     public void testBackendUsesRegisteredKryoDefaultSerializer() throws Exception {
```
```suggestion
    void testBackendUsesRegisteredKryoDefaultSerializer() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -570,14 +555,14 @@ public void testBackendUsesRegisteredKryoDefaultSerializer() throws Exception {
-    @Test
+    @TestTemplate
     @SuppressWarnings("unchecked")
     public void testBackendUsesRegisteredKryoDefaultSerializerUsingGetOrCreate() throws Exception {
```
```suggestion
    void testBackendUsesRegisteredKryoDefaultSerializerUsingGetOrCreate() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -644,14 +629,14 @@ public void testBackendUsesRegisteredKryoDefaultSerializerUsingGetOrCreate() thr
 
-    @Test
+    @TestTemplate
     public void testBackendUsesRegisteredKryoSerializer() throws Exception {
```
```suggestion
    void testBackendUsesRegisteredKryoSerializer() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -711,14 +697,14 @@ public void testBackendUsesRegisteredKryoSerializer() throws Exception {
-    @Test
+    @TestTemplate
     @SuppressWarnings("unchecked")
     public void testBackendUsesRegisteredKryoSerializerUsingGetOrCreate() throws Exception {
```
```suggestion
    void testBackendUsesRegisteredKryoSerializerUsingGetOrCreate() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -796,15 +783,16 @@ public void testBackendUsesRegisteredKryoSerializerUsingGetOrCreate() throws Exc
      */
-    @Test
+    @TestTemplate
     public void testKryoRegisteringRestoreResilienceWithRegisteredType() throws Exception {
```
```suggestion
    void testKryoRegisteringRestoreResilienceWithRegisteredType() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -872,7 +860,7 @@ public void testKryoRegisteringRestoreResilienceWithRegisteredType() throws Exce
-    @Test
+    @TestTemplate
     @SuppressWarnings("unchecked")
     public void testKryoRegisteringRestoreResilienceWithDefaultSerializer() throws Exception {
```
```suggestion
    void testKryoRegisteringRestoreResilienceWithDefaultSerializer() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -996,7 +1004,7 @@ public void testKryoRegisteringRestoreResilienceWithDefaultSerializer() throws E
      */
-    @Test
+    @TestTemplate
     public void testKryoRegisteringRestoreResilienceWithRegisteredSerializer() throws Exception {
```
```suggestion
    void testKryoRegisteringRestoreResilienceWithRegisteredSerializer() throws Exception {
```
could be package private

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -959,24 +947,44 @@ public void testKryoRegisteringRestoreResilienceWithDefaultSerializer() throws E
+                                            .isInstanceOf(ExpectedKryoTestException.class),
+                            actual ->
+                                    assertThat(actual)
+                                            .hasFieldOrProperty("cause")
```
Could you please clarify why should we check it?
In my opinion `Throwable` has `cause` field and this check will always pass for anything extending `Throwable`

jiexray replied:
This is because some catch blocks wrap the inner exception, i.e., `ExpectedKryoTestException` (thrown by `CustomKryoTestSerializer`).

For example, `HeapKeyedStateBackendBuilder#restoreState` wrap `ExpectedKryoTestException` in `BackendBuildingException`.

Thus, we need to check the `cause` field of the thrown exception.

snuyanzin replied:
there is `.hasCauseInstanceOf` method in assertJ e.g.
```java
assertThatThrownBy(() -> ...)
.hasCauseInstanceOf(ExpectedCauseException.class);
```
will not it help here?

jiexray replied:
Yes, I think `hasCauseInstanceOf()` is sufficient for the assertion, and `.hasFieldOrProperty("cause")` can be removed. The assertion code will become:

```
.satisfiesAnyOf(
        actual ->
                assertThat(actual)
                        .isInstanceOf(ExpectedKryoTestException.class),
        actual ->
                assertThat(actual)
                        .hasCauseInstanceOf(ExpectedKryoTestException.class));
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -1080,23 +1088,41 @@ public void testKryoRegisteringRestoreResilienceWithRegisteredSerializer() throw
+                                            .isInstanceOf(ExpectedKryoTestException.class),
+                            actual ->
+                                    assertThat(actual)
+                                            .hasFieldOrProperty("cause")
```
Could you please clarify why should we check it?
In my opinion `Throwable` has `cause` field and this check will always pass for anything extending `Throwable`

jiexray replied:
The same condition as mentioned in the former question.

jiexray commented:
Thank you @snuyanzin for your comments, I will fix them as soon as possible.

jiexray commented:
Hello @snuyanzin, I have fixed your comments with a new commit. The modifications includes:

- Remove public statement from test function
- Replace `ExceptionUtils.findThrowable` with `assertThatThrownBy`
- Replace `TestLogger` in junit4 with `TestLoggerExtension` in junit5

Myasuka commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/FileStateBackendMigrationTest.java:
```diff
@@ -18,18 +18,24 @@
 
+    @Parameters
+    public static Collection<Object> modes() {
+        return Arrays.asList("just to please junit5 parameter test in base class");
```
I don't think this sentence could be understood well.

jiexray replied:
I have changed the expression in the parameter list to enhanace reabability.

```
    @Parameters
    public static Collection<Object> modes() {
        // current test does not need to be parameterized
        return Collections.singletonList("no-use");
    }
```

Myasuka commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendTest.java:
```diff
@@ -148,13 +145,16 @@ public static List<Object[]> modes() {
-        String dbPath = new File(TEMP_FOLDER.newFolder(), DB_INSTANCE_DIR_STRING).getAbsolutePath();
+        File dbPath = new File(tmpDbPath, DB_INSTANCE_DIR_STRING);
+        if (!dbPath.exists()) {
+            dbPath.mkdirs();
```
Why we need this after switching junit5?

jiexray replied:
Thank you for pointing it out. In fact, this "manually creating path" is not required. The root problem is caused by the wrong "static" definition for field `tmpDbPath`.

Some background: "Static" in JUnit5 `@TempDir`  has special meaning. Static field means that the temporary directory is shared among **ALL** test methods. In contrast, non-static field means each test would use its own temporary directory.

In this utcase, each test should have an exclusive temporary directory for db. Thus, I change `tmpDbPath` to non-static field, and remove the "manually creating path". Every test is passed.

By the way, the checkpoint path (`tmpCheckpointPath`) is shared by all tests.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -61,17 +62,7 @@
-                return entry.getValue();
-            }
-
+        return new AbstractMap.SimpleEntry<UK, UV>(entry.getKey(), entry.getValue()) {
```
I don't this change has relationship with junit5 refactor.

Myasuka reviewed (commented):
Thanks for creating this PR, please take a look at my comments.

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:
```diff
@@ -152,17 +137,16 @@
  */
 @SuppressWarnings("serial")
-public abstract class StateBackendTestBase<B extends AbstractStateBackend> extends TestLogger {
+@ExtendWith({ParameterizedTestExtension.class, TestLoggerExtension.class})
```
No, I mean not `TestLoggerExtension`
Using the `META-INF/services/org.junit.jupiter.api.extension.Extension` is the better approach here.
for more details have a look at
https://github.com/apache/flink/commit/ecf67c67dbcf633c4533f8cb719f32512c19d0ce#diff-d96a3b8d09f526305595964a0010cf113e59999006534653bf45b9971b66241f

jiexray replied:
Thank you for your advice.

However, I find that `META-INF/services/org.junit.jupiter.api.extension.Extension` already exists in `flink-runtime/src/test/resources/` directory, and the content in the file is:
```
org.apache.flink.util.TestLoggerExtension
```

I think the `TestLoggerExtension.class` I used here is the `org.apache.flink.util.TestLoggerExtension` in the `flink-runtime/src/test/resources/META-INF/services/org.junit.jupiter.api.extension.Extension`

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateEmbeddedRocksDBStateBackendTest.java:
```diff
@@ -33,22 +33,22 @@
 import java.io.IOException;
 
 /** Tests for {@link ChangelogStateBackend} delegating {@link EmbeddedRocksDBStateBackend}. */
 public class ChangelogDelegateEmbeddedRocksDBStateBackendTest
```
```suggestion
class ChangelogDelegateEmbeddedRocksDBStateBackendTest
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateFileStateBackendTest.java:
```diff
@@ -35,20 +35,20 @@
 import java.io.IOException;
 
 /** Tests for {@link ChangelogStateBackend} delegating {@link FsStateBackend}. */
 public class ChangelogDelegateFileStateBackendTest extends FileStateBackendTest {
```
```suggestion
class ChangelogDelegateFileStateBackendTest extends FileStateBackendTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateHashMapInMemoryTest.java:
```diff
@@ -22,17 +22,12 @@
  * Tests for {@link ChangelogStateBackend} using {@link InMemoryStateChangelogStorage} and
  * delegating {@link HashMapStateBackendTest}.
  */
 public class ChangelogDelegateHashMapInMemoryTest extends ChangelogDelegateHashMapTest {
```
```suggestion
class ChangelogDelegateHashMapInMemoryTest extends ChangelogDelegateHashMapTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateHashMapTest.java:
```diff
@@ -32,19 +32,19 @@
 import java.io.IOException;
 
 /** Tests for {@link ChangelogStateBackend} delegating {@link HashMapStateBackendTest}. */
 public class ChangelogDelegateHashMapTest extends HashMapStateBackendTest {
```
```suggestion
class ChangelogDelegateHashMapTest extends HashMapStateBackendTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateMemoryStateBackendTest.java:
```diff
@@ -35,20 +35,20 @@
 import java.io.IOException;
 
 /** Tests for {@link ChangelogStateBackend} delegating {@link MemoryStateBackend}. */
 public class ChangelogDelegateMemoryStateBackendTest extends MemoryStateBackendTest {
```
```suggestion
class ChangelogDelegateMemoryStateBackendTest extends MemoryStateBackendTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateStateTest.java:
```diff
@@ -32,74 +32,77 @@
 /** Tests for {@link ChangelogStateBackend} delegating state accesses. */
-@RunWith(Parameterized.class)
+@ExtendWith(ParameterizedTestExtension.class)
 public class ChangelogDelegateStateTest {
```
```suggestion
class ChangelogDelegateStateTest {
```
could be package private

jiexray replied:
`@ExtendWith(ParameterizedTestExtension.class)` cannot use package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackendTest.java:
```diff
@@ -38,22 +38,23 @@
 /** {@link ChangelogKeyedStateBackend} test. */
-@RunWith(Parameterized.class)
+@ExtendWith(ParameterizedTestExtension.class)
 public class ChangelogKeyedStateBackendTest {
```
```suggestion
class ChangelogKeyedStateBackendTest {
```
could be package private

jiexray replied:
`@ExtendWith(ParameterizedTestExtension.class)` cannot use package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogListStateTest.java:
```diff
@@ -38,50 +38,48 @@
 
 /** ChangelogListState Test. */
 @SuppressWarnings({"rawtypes", "unchecked"})
 public class ChangelogListStateTest {
```
```suggestion
class ChangelogListStateTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogMapStateTest.java:
```diff
@@ -36,80 +36,80 @@
 
 /** ChangelogMapState Test. */
 @SuppressWarnings({"rawtypes", "unchecked"})
 public class ChangelogMapStateTest {
```
```suggestion
class ChangelogMapStateTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogPqStateTest.java:
```diff
@@ -40,57 +40,55 @@
 
 /** ChangelogKeyGroupedPriorityQueue Test. */
 @SuppressWarnings({"rawtypes", "unchecked"})
 public class ChangelogPqStateTest {
```
```suggestion
class ChangelogPqStateTest {
```
could be package private

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java:
```diff
@@ -58,75 +58,73 @@
                 CheckpointStorageLoader.load(null, null, backend, config(), cl, null);
-
-        assertTrue(backend instanceof HashMapStateBackend);
+        assertThat(backend instanceof HashMapStateBackend).isTrue();
```
```suggestion
        assertThat(backend).isInstanceOf(HashMapStateBackend.class);
```
could be simplified

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java:
```diff
@@ -229,14 +234,14 @@ public void testLoadingRocksDBStateBackend() throws Exception {
         // set back and force
         env.setStateBackend(new MemoryStateBackend());
-        assertTrue(env.getStateBackend() instanceof MemoryStateBackend);
+        assertThat(env.getStateBackend() instanceof MemoryStateBackend).isTrue();
```
```suggestion
        assertThat(env.getStateBackend()).isInstanceOf(MemoryStateBackend.class);
```
could be simplified

snuyanzin commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendMigrationTest.java:
```diff
@@ -24,28 +24,26 @@
 
 /** Tests for the partitioned state part of {@link ChangelogStateBackend}. */
-@RunWith(Parameterized.class)
 public class ChangelogStateBackendMigrationTest
```
```suggestion
class ChangelogStateBackendMigrationTest
```
could be package private

jiexray replied:
`@ExtendWith(ParameterizedTestExtension.class)` cannot use package private

snuyanzin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendTest.java:
```diff
@@ -453,12 +452,12 @@ public void testCompletingSnapshot() throws Exception {
-            assertTrue(keyedStateHandle.getStateSize() > 0);
-            assertEquals(2, keyedStateHandle.getKeyGroupRange().getNumberOfKeyGroups());
+            assertThat(keyedStateHandle).isNotNull();
+            assertThat(keyedStateHandle.getStateSize() > 0).isTrue();
```
```suggestion
            assertThat(keyedStateHandle.getStateSize()).isGreaterThan(0);
```
could be simplified

snuyanzin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendTest.java:
```diff
@@ -529,7 +528,7 @@ public void testDisposeDeletesAllDirectories() throws Exception {
 
             // more than just the root directory
-            assertTrue(allFilesInDbDir.size() > 1);
+            assertThat(allFilesInDbDir.size() > 1).isTrue();
```
```suggestion
            assertThat(allFilesInDbDir).hasSizeGreaterThan(1);
```
could be simplified

jiexray commented:
@Myasuka Thank you for your comments. I have fixed them in a new commit. https://github.com/apache/flink/pull/21410/commits/bf87b002395acf1c6d0a606f6699544087769bd7

jiexray commented:
@roncohen Thank you for you comments. I have fixed some problems in the commit https://github.com/apache/flink/pull/21410/commits/944bf2a8fe21628d279df006fbc417934a122381.

However, I am still confused about the problem related to `TestLoggerExtension`. I have replied it just after your comment.

jiexray commented:
@flinkbot run azure

Myasuka commented:
@jiexray Please click `Resolve converasation` button if you think you already addressed the comment.

jiexray commented:
> @jiexray Please click `Resolve converasation` button if you think you already addressed the comment.

OK, I will re-check all converasation, and click `Resolve converasation`.

snuyanzin commented:
Hi @jiexray I wonder whether you would like to continue working on this PR or not

jiexray commented:
> Hi @jiexray I wonder whether you would like to continue working on this PR or not

Thank you for you reminding. I have resolved all converasation, and I will rebase my pr this week.

jiexray commented:
@flinkbot  run azure

jiexray commented:
@flinkbot  run azure

jiexray commented:
@Myasuka @fredia I have resolved all conversations? Could you have a look at this pr again.

masteryhx commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -83,6 +84,16 @@ public UV setValue(UV value) {
             }
+
+            @Override
+            public boolean equals(Object o) {
```
This fix should not be related to Junit5 migration, right?
It's better to seprate them into 2 commits.

masteryhx reviewed (changes requested):
Thanks for the update. Overall LGTM.
Please squash and seprate some commits.

fredia commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendMigrationTest.java:
```diff
@@ -21,33 +21,37 @@
 public class RocksDBStateBackendMigrationTest
         extends StateBackendMigrationTestBase<RocksDBStateBackend> {
+    // Store it because we need it for the cleanup test.
+    public static String dbPath;
```
why mark `dbPath` as `public static`?

fredia commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendMigrationTest.java:
```diff
@@ -22,25 +22,30 @@
 
 /** Tests for the partitioned state part of {@link EmbeddedRocksDBStateBackend}. */
-@RunWith(Parameterized.class)
 public class EmbeddedRocksDBStateBackendMigrationTest
```
Remove `public`?

fredia commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java:
```diff
@@ -58,75 +58,73 @@
-import static org.junit.Assert.assertTrue;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertSame;
```
Can replace `assertSame` with assertj?

fredia commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/FileStateBackendMigrationTest.java:
```diff
@@ -18,18 +18,25 @@
  * Tests for the keyed state backend and operator state backend, as created by the {@link
  * FsStateBackend}.
  */
 public class FileStateBackendMigrationTest extends StateBackendMigrationTestBase<FsStateBackend> {
```
Remove `public`?

fredia reviewed (changes requested):
Thanks for the update, It would be better to remove the Mockito according to code style:
https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#avoid-mockito---use-reusable-test-implementations

Would you like to remove the Mockito in `EmbeddedRocksDBStateBackendTest`? This is out of the scope of the `flink-statebackend-changelog` module, maybe we can do this in another PR.

jiexray commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendTest.java:
```diff
@@ -46,18 +46,17 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.io.TempDir;
 import org.mockito.Mockito;
```
Good point. I think this would better be changed in another pr.

masteryhx reviewed (changes requested):
@jiexray Hi, Could you rebase your pr with master ?
Seems there are lots of conflicts due to [FLINK-32854](https://issues.apache.org/jira/browse/FLINK-32854)
